File: android/app/src/main/java/io/github/shadowur0/netsira/devices/AirOsClient.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
package io.github.shadowur0.netsira.devices

import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.FormBody
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.net.URI
import java.util.concurrent.TimeUnit

data class AirOsSnapshot(
    val apiVersion: Int,
    val model: String,
    val hostname: String?,
    val firmware: String?,
    val signalDbm: Double?,
    val noiseDbm: Double?,
    val snrDb: Double?,
    val chainRssi: String?,
    val frequencyMHz: Double?,
    val channelWidthMHz: Double?,
    val txPowerDbm: Double?,
    val distanceMeters: Double?,
    val txRate: String?,
    val rxRate: String?,
    val ethernetSpeedMbps: Double?,
    val ethernetFullDuplex: Boolean?,
    val cpuLoadPercent: Double?
)

class AirOsClient {
    private class MemoryCookieJar : CookieJar {
        private val cookies = mutableMapOf<String, List<Cookie>>()
        override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
            this.cookies[url.host] = cookies
        }
        override fun loadForRequest(url: HttpUrl): List<Cookie> = cookies[url.host].orEmpty()
    }

    private val client = OkHttpClient.Builder()
        .cookieJar(MemoryCookieJar())
        .followRedirects(false)
        .followSslRedirects(false)
        .connectTimeout(8, TimeUnit.SECONDS)
        .readTimeout(8, TimeUnit.SECONDS)
        .build()

    private var csrf: String? = null

    fun connectAndRead(baseValue: String, username: String, password: String): AirOsSnapshot {
        val base = normalize(baseValue)
        val api = if (tryLoginV8(base, username, password)) 8 else loginV6(base, username, password)

        val request = Request.Builder()
            .url(resolve(base, "status.cgi"))
            .header("User-Agent", "Netsira/0.2")
            .apply { csrf?.let { header("X-CSRF-ID", it) } }
            .build()

        client.newCall(request).execute().use { response ->
            if (response.code == 401 || response.code == 403) error("airOS rejected the authenticated status request.")
            if (!response.isSuccessful) error("airOS status failed: HTTP " + response.code)
            return parseSnapshot(api, JSONObject(response.body.string()))
        }
    }

    private fun tryLoginV8(base: String, username: String, password: String): Boolean {
        val json = JSONObject().put("username", username).put("password", password).toString()
        val request = Request.Builder()
            .url(resolve(base, "api/auth"))
            .header("User-Agent", "Netsira/0.2")
            .post(json.toRequestBody("application/json".toMediaType()))
            .build()

        client.newCall(request).execute().use { response ->
            if (response.code == 404) return false
            if (response.code == 401 || response.code == 403) error("Invalid airOS username or password.")
            if (!response.isSuccessful) error("airOS 8 login failed: HTTP " + response.code)
            csrf = response.header("X-CSRF-ID")
            return true
        }
    }

    private fun loginV6(base: String, username: String, password: String): Int {
        val loginUrl = resolve(base, "login.cgi")
        client.newCall(Request.Builder().url(loginUrl).get().build()).execute().use {
            if (!it.isSuccessful) error("airOS 6 login page failed: HTTP " + it.code)
        }

        val form = FormBody.Builder()
            .add("username", username)
            .add("password", password)
            .add("uri", "/index.cgi")
            .build()
        val post = Request.Builder()
            .url(loginUrl)
            .header("Referer", loginUrl)
            .header("Origin", URI(base).let { it.scheme + "://" + it.authority })
            .post(form)
            .build()

        client.newCall(post).execute().use { response ->
            if (response.code !in listOf(301, 302, 303, 307, 308)) error("airOS 6 login failed.")
        }

        client.newCall(Request.Builder().url(resolve(base, "index.cgi")).get().build()).execute().use { response ->
            val location = response.header("Location").orEmpty()
            if (response.code in listOf(301, 302, 303, 307, 308) && location.contains("login.cgi", ignoreCase = true)) {
                error("airOS 6 session activation failed.")
            }
        }
        return 6
    }

    private fun parseSnapshot(api: Int, root: JSONObject): AirOsSnapshot {
        val host = root.optJSONObject("host")
        val wireless = root.optJSONObject("wireless")
        val station = wireless?.optJSONArray("sta")?.optJSONObject(0)
        val remote = station?.optJSONObject("remote")

        val signal = number(wireless, "signal") ?: number(station, "signal") ?: number(remote, "signal")
        val noise = number(wireless, "noisef") ?: number(station, "noisefloor") ?: number(remote, "noisefloor")
        val chains = arrayNumbers(station?.optJSONArray("chainrssi")) ?: arrayNumbers(remote?.optJSONArray("chainrssi"))
        val ethernet = firstInterfaceStatus(root.optJSONArray("interfaces"))

        return AirOsSnapshot(
            apiVersion = api,
            model = text(host, "devmodel") ?: "Unknown",
            hostname = text(host, "hostname"),
            firmware = text(host, "fwversion"),
            signalDbm = signal,
            noiseDbm = noise,
            snrDb = if (signal != null && noise != null) signal - noise else null,
            chainRssi = chains,
            frequencyMHz = number(wireless, "frequency"),
            channelWidthMHz = number(wireless, "chanbw"),
            txPowerDbm = number(wireless, "txpower"),
            distanceMeters = number(wireless, "distance") ?: number(station, "distance") ?: number(remote, "distance"),
            txRate = text(wireless, "txrate"),
            rxRate = text(wireless, "rxrate"),
            ethernetSpeedMbps = number(ethernet, "speed"),
            ethernetFullDuplex = bool(ethernet, "duplex"),
            cpuLoadPercent = number(host, "cpuload")
        )
    }

    private fun firstInterfaceStatus(array: JSONArray?): JSONObject? {
        if (array == null) return null
        for (i in 0 until array.length()) {
            array.optJSONObject(i)?.optJSONObject("status")?.let { return it }
        }
        return null
    }

    private fun number(obj: JSONObject?, key: String): Double? {
        if (obj == null || !obj.has(key) || obj.isNull(key)) return null
        val value = obj.opt(key)
        return when (value) {
            is Number -> value.toDouble()
            is String -> value.substringBefore(' ').toDoubleOrNull()
            else -> null
        }
    }

    private fun text(obj: JSONObject?, key: String): String? =
        obj?.optString(key)?.takeIf { it.isNotBlank() && it != "null" }

    private fun bool(obj: JSONObject?, key: String): Boolean? {
        if (obj == null || !obj.has(key) || obj.isNull(key)) return null
        return when (val value = obj.opt(key)) {
            is Boolean -> value
            is Number -> value.toInt() != 0
            else -> null
        }
    }

    private fun arrayNumbers(array: JSONArray?): String? {
        if (array == null || array.length() == 0) return null
        return (0 until array.length()).joinToString(", ") { array.opt(it).toString() }
    }

    private fun normalize(value: String): String {
        var candidate = value.trim()
        if (!candidate.contains("://")) candidate = "http://" + candidate
        val uri = URI(candidate)
        require(uri.scheme == "http" || uri.scheme == "https") { "Only HTTP or HTTPS is supported." }
        require(!uri.host.isNullOrBlank()) { "Invalid device address." }
        return URI(uri.scheme, uri.userInfo, uri.host, uri.port, "/", null, null).toString()
    }

    private fun resolve(base: String, path: String): String = URI(base).resolve(path).toString()
}

File: desktop/Netsira.Desktop/Devices/AirOsClient.cs
// SPDX-License-Identifier: AGPL-3.0-or-later
using System.Net;
using System.Text;
using System.Text.Json;

namespace Netsira.Desktop.Devices;

public sealed record AirOsSnapshot(
    int ApiVersion,
    string Model,
    string? Hostname,
    string? Firmware,
    double? SignalDbm,
    double? NoiseDbm,
    double? SnrDb,
    string? ChainRssi,
    double? FrequencyMHz,
    double? ChannelWidthMHz,
    double? TxPowerDbm,
    double? DistanceMeters,
    string? TxRate,
    string? RxRate,
    double? EthernetSpeedMbps,
    bool? EthernetFullDuplex,
    double? CpuLoadPercent,
    long? TotalRamBytes,
    long? FreeRamBytes);

public sealed class AirOsClient : IDisposable
{
    private readonly HttpClient _http;
    private string? _csrf;
    private Uri? _baseUri;
    private int? _apiVersion;

    public AirOsClient(bool allowInvalidCertificate = false)
    {
        var handler = new HttpClientHandler
        {
            AllowAutoRedirect = false,
            CookieContainer = new CookieContainer(),
            UseCookies = true
        };
        if (allowInvalidCertificate)
            handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;

        _http = new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(10) };
        _http.DefaultRequestHeaders.UserAgent.ParseAdd("Netsira/0.3");
    }

    public bool IsAuthenticated => _baseUri is not null && _apiVersion is not null;

    public async Task LoginAsync(
        string baseUrl,
        string username,
        string password,
        CancellationToken cancellationToken = default)
    {
        _baseUri = NormalizeBaseUri(baseUrl);
        _csrf = null;
        _apiVersion = await TryLoginV8Async(_baseUri, username, password, cancellationToken)
            ? 8
            : await LoginV6Async(_baseUri, username, password, cancellationToken);
    }

    public async Task<AirOsSnapshot> ReadStatusAsync(CancellationToken cancellationToken = default)
    {
        if (_baseUri is null || _apiVersion is null)
            throw new InvalidOperationException("airOS client is not authenticated.");

        using var request = new HttpRequestMessage(HttpMethod.Get, new Uri(_baseUri, "status.cgi"));
        if (!string.IsNullOrWhiteSpace(_csrf))
            request.Headers.TryAddWithoutValidation("X-CSRF-ID", _csrf);

        using var response = await _http.SendAsync(request, cancellationToken);
        if (response.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
            throw new UnauthorizedAccessException("airOS session expired or status access was rejected.");
        response.EnsureSuccessStatusCode();

        await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken);
        using var doc = await JsonDocument.ParseAsync(stream, cancellationToken: cancellationToken);
        return ParseSnapshot(_apiVersion.Value, doc.RootElement);
    }

    public async Task<AirOsSnapshot> ConnectAndReadAsync(
        string baseUrl,
        string username,
        string password,
        CancellationToken cancellationToken = default)
    {
        await LoginAsync(baseUrl, username, password, cancellationToken);
        return await ReadStatusAsync(cancellationToken);
    }

    private async Task<bool> TryLoginV8Async(Uri baseUri, string username, string password, CancellationToken ct)
    {
        var payload = "{\"username\":\"" + JsonEncodedText.Encode(username) +
            "\",\"password\":\"" + JsonEncodedText.Encode(password) + "\"}";
        using var response = await _http.PostAsync(
            new Uri(baseUri, "api/auth"),
            new StringContent(payload, Encoding.UTF8, "application/json"),
            ct);

        if (response.StatusCode == HttpStatusCode.NotFound) return false;
        if (response.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
            throw new UnauthorizedAccessException("Invalid airOS username or password.");

        response.EnsureSuccessStatusCode();
        _csrf = response.Headers.TryGetValues("X-CSRF-ID", out var values) ? values.FirstOrDefault() : null;
        return true;
    }

    private async Task<int> LoginV6Async(Uri baseUri, string username, string password, CancellationToken ct)
    {
        var loginUri = new Uri(baseUri, "login.cgi");
        using (var initial = await _http.GetAsync(loginUri, ct))
            initial.EnsureSuccessStatusCode();

        using var post = new HttpRequestMessage(HttpMethod.Post, loginUri)
        {
            Content = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["username"] = username,
                ["password"] = password,
                ["uri"] = "/index.cgi"
            })
        };
        post.Headers.Referrer = loginUri;
        post.Headers.TryAddWithoutValidation("Origin", baseUri.GetLeftPart(UriPartial.Authority));

        using var login = await _http.SendAsync(post, ct);
        if ((int)login.StatusCode is not (301 or 302 or 303 or 307 or 308))
            throw new UnauthorizedAccessException("airOS 6 login failed.");

        using var activate = await _http.GetAsync(new Uri(baseUri, "index.cgi"), ct);
        if (IsLoginRedirect(activate))
            throw new UnauthorizedAccessException("airOS 6 session activation failed.");
        if (!activate.IsSuccessStatusCode && !IsRedirect(activate.StatusCode))
            activate.EnsureSuccessStatusCode();

        return 6;
    }

    private static bool IsRedirect(HttpStatusCode code) => (int)code is 301 or 302 or 303 or 307 or 308;

    private static bool IsLoginRedirect(HttpResponseMessage response) =>
        IsRedirect(response.StatusCode) &&
        response.Headers.Location?.ToString().Contains("login.cgi", StringComparison.OrdinalIgnoreCase) == true;

    private static Uri NormalizeBaseUri(string value)
    {
        var trimmed = value.Trim();
        if (!trimmed.Contains("://", StringComparison.Ordinal)) trimmed = "http://" + trimmed;
        var uri = new Uri(trimmed, UriKind.Absolute);
        if (uri.Scheme is not ("http" or "https")) throw new ArgumentException("Only HTTP or HTTPS is supported.");
        var builder = new UriBuilder(uri) { Path = "/", Query = "", Fragment = "" };
        return builder.Uri;
    }

    private static AirOsSnapshot ParseSnapshot(int apiVersion, JsonElement root)
    {
        TryGet(root, out var host, "host");
        TryGet(root, out var wireless, "wireless");

        var station = FirstArrayItem(wireless, "sta");
        var remote = station.HasValue && TryGet(station.Value, out var remoteValue, "remote") ? remoteValue : (JsonElement?)null;

        var signal = Number(wireless, "signal") ?? Number(station, "signal") ?? Number(remote, "signal");
        var noise = Number(wireless, "noisef") ?? Number(station, "noisefloor") ?? Number(remote, "noisefloor");
        double? snr = signal.HasValue && noise.HasValue ? signal.Value - noise.Value : null;

        var chain = ArrayNumbers(station, "chainrssi") ?? ArrayNumbers(remote, "chainrssi");
        var ethernet = FirstInterfaceWithStatus(root);

        return new AirOsSnapshot(
            ApiVersion: apiVersion,
            Model: Text(host, "devmodel") ?? "Unknown",
            Hostname: Text(host, "hostname"),
            Firmware: Text(host, "fwversion"),
            SignalDbm: signal,
            NoiseDbm: noise,
            SnrDb: snr,
            ChainRssi: chain,
            FrequencyMHz: Number(wireless, "frequency"),
            ChannelWidthMHz: Number(wireless, "chanbw"),
            TxPowerDbm: Number(wireless, "txpower"),
            DistanceMeters: Number(wireless, "distance") ?? Number(station, "distance") ?? Number(remote, "distance"),
            TxRate: Text(wireless, "txrate"),
            RxRate: Text(wireless, "rxrate"),
            EthernetSpeedMbps: Number(ethernet, "speed"),
            EthernetFullDuplex: Boolean(ethernet, "duplex"),
            CpuLoadPercent: Number(host, "cpuload"),
            TotalRamBytes: Integer(host, "totalram"),
            FreeRamBytes: Integer(host, "freeram"));
    }

    private static JsonElement? FirstInterfaceWithStatus(JsonElement root)
    {
        if (!TryGet(root, out var interfaces, "interfaces") || interfaces.ValueKind != JsonValueKind.Array) return null;
        foreach (var item in interfaces.EnumerateArray())
        {
            if (TryGet(item, out var status, "status") && status.ValueKind == JsonValueKind.Object)
                return status;
        }
        return null;
    }

    private static JsonElement? FirstArrayItem(JsonElement? parent, string name)
    {
        if (!parent.HasValue || !TryGet(parent.Value, out var array, name) || array.ValueKind != JsonValueKind.Array) return null;
        foreach (var item in array.EnumerateArray()) return item;
        return null;
    }

    private static bool TryGet(JsonElement root, out JsonElement value, string name)
    {
        if (root.ValueKind == JsonValueKind.Object && root.TryGetProperty(name, out value)) return true;
        value = default;
        return false;
    }

    private static double? Number(JsonElement? parent, string name)
    {
        if (!parent.HasValue || !TryGet(parent.Value, out var value, name)) return null;
        if (value.ValueKind == JsonValueKind.Number && value.TryGetDouble(out var number)) return number;
        if (value.ValueKind == JsonValueKind.String && double.TryParse(value.GetString()?.Split(' ')[0], out number)) return number;
        return null;
    }

    private static long? Integer(JsonElement? parent, string name)
    {
        if (!parent.HasValue || !TryGet(parent.Value, out var value, name)) return null;
        if (value.ValueKind == JsonValueKind.Number && value.TryGetInt64(out var number)) return number;
        if (value.ValueKind == JsonValueKind.String && long.TryParse(value.GetString()?.Split(' ')[0], out number)) return number;
        return null;
    }

    private static string? Text(JsonElement? parent, string name)
    {
        if (!parent.HasValue || !TryGet(parent.Value, out var value, name)) return null;
        return value.ValueKind == JsonValueKind.String ? value.GetString() : value.ToString();
    }

    private static bool? Boolean(JsonElement? parent, string name)
    {
        if (!parent.HasValue || !TryGet(parent.Value, out var value, name)) return null;
        return value.ValueKind switch
        {
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.Number when value.TryGetInt32(out var n) => n != 0,
            _ => null
        };
    }

    private static string? ArrayNumbers(JsonElement? parent, string name)
    {
        if (!parent.HasValue || !TryGet(parent.Value, out var value, name) || value.ValueKind != JsonValueKind.Array) return null;
        return string.Join(", ", value.EnumerateArray().Select(v => v.ToString()));
    }

    public void Dispose() => _http.Dispose();
}
